### PR TITLE
Fix various HVACTemplate issues

### LINKED ISSUE #9018: HVACTemplate errors
Issue overview
--------------
More issues found in ExpandObjects while testing the new pyExpandObjects.

- [x] HVACTemplate:System:ConstantVolume: 
  * heating coil type should default to HotWater, but defaults to None
  * economizer type defaults to None but should be NoEconomizer (None causes fatal error).
[SystemConstantVolume_minimum_inputs.idf.txt](https://github.com/NREL/EnergyPlus/files/7068692/SystemConstantVolume_minimum_inputs.idf.txt)

- [x] HVACTemplate:System:PackagedVAV:
  * Minimal input with all defaults has a fatal error 
  `** Severe  ** <root>[ZoneControl:Humidistat][DXVAV Sys 1 Humidification Humidistat] - Missing required property 'zone_name'.`
  [PackagedVAV-minimum_inputs.idf.txt](https://github.com/NREL/EnergyPlus/files/7068715/PackagedVAV-minimum_inputs.idf.txt)
  
  * False warning even though heating setpoint is < cooling setpoint.
  ```
   **   ~~~   ** Warning:  In HVACTemplate:System:PackagedVAV "DXVAV Sys 1" the Heating Coil
   **   ~~~   ** Design Setpoint is greater than the Cooling Coil Design Setpoint. This may
  ```
  [PackagedVAV-heating_coil_design_setpoint.idf.txt](https://github.com/NREL/EnergyPlus/files/7068729/PackagedVAV-heating_coil_design_setpoint.idf.txt)

- [x] HVACTemplate:System:Unitary
  * Expanded SetpointManager:SingleZone:Cooling "Minimum Supply Air Temperature" does not change with template input for Cooling Coil Design Supply Air Temperature.
  [SystemUnitary-cooling_design_supply_air_temperature.idf.txt](https://github.com/NREL/EnergyPlus/files/7068857/SystemUnitary-cooling_design_supply_air_temperature.idf.txt)

  * Expanded Controller:OutdoorAir "Economizer Maximum Limit Dewpoint Temperature" should be applied regardless of economizer type.
  [SystemUnitary-economizer_maximum_limit_dewpoint_temperature.idf.txt](https://github.com/NREL/EnergyPlus/files/7068860/SystemUnitary-economizer_maximum_limit_dewpoint_temperature.idf.txt)

- [x] HVACTemplate:System:UnitaryHeatPump:AirToAir
  * Rated Air Flow Rate in the cooling and heating coils should always be autosize, not the template input for Cooling or Heating Supply Air Flow Rate. The entered flow rate goes into Sizing:System.
  [SystemUnitaryHeatPump-cooling_supply_air_flow_rate.idf.txt](https://github.com/NREL/EnergyPlus/files/7068920/SystemUnitaryHeatPump-cooling_supply_air_flow_rate.idf.txt)

- [x] HVACTemplate:System:VAV
  * economizer type defaults to None but should be NoEconomizer (None causes fatal error).
  [SystemVAV-minimum_inputs.idf.txt](https://github.com/NREL/EnergyPlus/files/7069035/SystemVAV-minimum_inputs.idf.txt)

  * System with no preheat coil and no heating coil produces two warnings about heating setpoint, only the second one is relevant for this case.
  ```
   ** Warning ** Output:PreprocessorMessage="ExpandObjects" has the following Warning conditions:
   **   ~~~   ** Warning:  In HVACTemplate:System:VAV "VAV Sys 1" the Heating Coil Design
   **   ~~~   ** Setpoint is greater than the Preheat Coil Design Setpoint
   **   ~~~   ** but Heating Coil
   **   ~~~   ** Type=None. Using Preheat Coil Design Setpoint for the Sizing:System Central
   **   ~~~   ** Heating Design Supply Air Temperature.
   ** Warning ** Output:PreprocessorMessage="ExpandObjects" has the following Warning conditions:
   **   ~~~   ** Warning:  In HVACTemplate:System:VAV "VAV Sys 1" there is no Heating Coil
   **   ~~~   ** and no Preheat Coil. The Preheat Coil Design Setpoint will be used for the
   **   ~~~   ** Sizing:System Central Heating Design Supply Air Temperature. This will be
   **   ~~~   ** the inlet air temperature for sizing reheat coils.
  ```
  [SystemVAV-heating_coil_type_none.idf.txt](https://github.com/NREL/EnergyPlus/files/7069039/SystemVAV-heating_coil_type_none.idf.txt)

  * System with preheat coil and heat recovery lists the components in the wrong order in the expanded AirLoopHVAC:OutdoorAirSystem:EquipmentList. The heat recovery should be first, then the preheat coil to be consistent with the airflow order of the node connections. 
  * The same problem was fixed for HVACTemplate:System:DualDuct (not test file made for that)
  [SystemVAV-preheat_coil_type_hot_water_heat_recovery_type_sensible.idf.txt](https://github.com/NREL/EnergyPlus/files/7069037/SystemVAV-preheat_coil_type_hot_water_heat_recovery_type_sensible.idf.txt)

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 9.6.0 I/O Freeze

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #9108: Use Enumerations Instead of Strings to Internally Refer to Types
```cpp
void FindAirPlantCondenserLoopFromBranchList(EnergyPlusData &state,
                                                 std::string const &BranchListName, // Branch List Name
                                                 std::string &LoopType,             // LoopType (if found, Plant,Condenser or Air)
                                                 std::string &LoopSupplyDemandAir,  // Supply if "Supply" or Demand if "Demand" or Air if "Air"
                                                 bool &MatchedLoop                  // true if found
    )
```

There are multiple issues with these design of this function, but the first one is that `LoopType` and `LoopSupplyDemandAir` should be enumerations and not strings. Strings are slow, and they are for people. They should only be used in input and output processing. Everything else should use indices and enumerations. A quick grep shows that there are 314 such function parameters in EnergyPlus.

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #9018 See the issue for more details and explanations.
 - Fix various HVACTemplate issues.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by rraustad** [comment]
FYI, transitioned a V92 user file using #8997.

    Conversion 9.2 => 9.3
     Starting new Transition.audit
     Will create new full IDFs
     Will create new IDF lines with units where applicable
     Will create new IDF lines leaving blank incoming fields as blank (no default fill)
     Processing IDF -- C:\EnergyPlusV9-6-0\PreProcess\IDFVersionUpdater\IDF Proposed model_updated_V920.idf
       ** Warning ** Object=Controller:MechanicalVentilation, name=Controller Mechanical Ventilation 3, entered with less than minimum number of fields.
       **   ~~~   ** Attempting fill to minimum.
       ** Severe  ** Object=Controller:MechanicalVentilation, name=Controller Mechanical Ventilation 3, Required Field=[Zone 1 Name] was blank.
       ** Warning ** Object=Controller:MechanicalVentilation, name=Controller Mechanical Ventilation 11, entered with less than minimum number of fields.
       **   ~~~   ** Attempting fill to minimum.
       ** Severe  ** Object=Controller:MechanicalVentilation, name=Controller Mechanical Ventilation 11, Required Field=[Zone 1 Name] was blank.
       ** Warning ** Object=Controller:MechanicalVentilation, name=Controller Mechanical Ventilation 63, entered with less than minimum number of fields.
       **   ~~~   ** Attempting fill to minimum.
       ** Severe  ** Object=Controller:MechanicalVentilation, name=Controller Mechanical Ventilation 63, Required Field=[Zone 1 Name] was blank.
       ** Severe  ** Possible incorrect IDD File
       **   ~~~   ** Possible Invalid Numerics or other problems

    Conversion 9.3 => 9.4
     Starting new Transition.audit
     Will create new full IDFs
     Will create new IDF lines with units where applicable
     Will create new IDF lines leaving blank incoming fields as blank (no default fill)
     Processing IDF -- C:\EnergyPlusV9-6-0\PreProcess\IDFVersionUpdater\IDF Proposed model_updated_V920.idf

    Conversion 9.4 => 9.5
     Starting new Transition.audit
     Will create new full IDFs
     Will create new IDF lines with units where applicable
     Will create new IDF lines leaving blank incoming fields as blank (no default fill)
     Processing IDF -- C:\EnergyPlusV9-6-0\PreProcess\IDFVersionUpdater\IDF Proposed model_updated_V920.idf

    Conversion 9.5 => 9.6
     Starting new Transition.audit
     Will create new full IDFs
     Will create new IDF lines with units where applicable
     Will create new IDF lines leaving blank incoming fields as blank (no default fill)
     Processing IDF -- C:\EnergyPlusV9-6-0\PreProcess\IDFVersionUpdater\IDF Proposed model_updated_V920.idf

**Comment by rraustad** [comment]
Then I see this warning when opening in IDFEditor.

image

**Comment by mjwitte** [comment]
Closing this for now. Will open a new PR when this is complete.

**Comment by mjwitte** [comment]
This fixes a collection of HVACTemplate issues. It's ready for review.

File: src/ExpandObjects/epfilter.f90 [hunk 1]
```diff
@@ -5684,7 +5684,7 @@ SUBROUTINE GetInputVAV
     CALL SetIfBlank(fldValStart + vsMaxOutsideFlowOff, 'autosize')
     CALL SetIfBlank(fldValStart + vsMinOutsideFlowOff, 'autosize')
     CALL SetIfBlank(fldValStart + vsMinOutsideControlOff, 'ProportionalMinimum')
-    CALL SetIfBlank(fldValStart + vsEconomizerKindOff, 'None')
+    CALL SetIfBlank(fldValStart + vsEconomizerKindOff, 'NoEconomizer')
```
Comment (Myoldmopar) [inline src/ExpandObjects/epfilter.f90:5687]: I can understand some of the changes here, but I don't have a feel for the full context of each change, or the impact/side effects.  Perhaps you could comment on some of them, but I also understand that these have stemmed from the vastly improved testing structure set up for the new Python expand objects tool.  So I am fine accepting them if you feel these are satisfying the issues identified over there.

File: src/ExpandObjects/epfilter.f90 [hunk 2]
```diff
@@ -14391,7 +14393,7 @@ SUBROUTINE CreateNewUnitary
   CALL CreateNewObj('SetpointManager:SingleZone:Cooling')
   CALL AddToObjFld('Name', base + usAirHandlerNameOff,' Economizer Supply Air Temp Manager')
   CALL AddToObjStr('Control Variable','Temperature')
-  CALL AddToObjStr('minimum supply air temperature {C}','13')
+  CALL AddToObjFld('minimum supply air temperature {C}',base + usCoolDesignSupplyTempOff,'')
```
Comment (Myoldmopar) [inline src/ExpandObjects/epfilter.f90:14396]: So the magic "13" is gone in this line, what about the magic "45" in the following line?

File: src/ExpandObjects/epfilter.f90 (hunk 1 above)
Comment (mjwitte) [reply src/ExpandObjects/epfilter.f90:5687]: @Myoldmopar There are more details in #9018 . Here the default wasn't even a valid choice for this object.

File: src/ExpandObjects/epfilter.f90 [hunk 3]
```diff
@@ -6171,7 +6171,7 @@ SUBROUTINE GetInputConstVol
     CALL SetIfBlank(fldValStart + cvsCoolResetODBLoOff, '15.6')
     CALL SetIfBlank(fldValStart + cvsCoolSetPtAtODBHiOff, '12.8')
     CALL SetIfBlank(fldValStart + cvsCoolResetODBHiOff, '23.3')
-    CALL SetIfBlank(fldValStart + cvsHeatCoilKindOff, 'None')
+    CALL SetIfBlank(fldValStart + cvsHeatCoilKindOff, 'HotWater')
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:6174]: Incorrect default fixed here.

File: src/ExpandObjects/epfilter.f90 [hunk 4]
```diff
@@ -10235,7 +10235,7 @@ SUBROUTINE CreateNewVAV
   ENDIF
   ! check if there is no heating coil and heating design setpoint is > preheat design setpoint
   IF (heatCoilType == ctNone) THEN
-    IF(StringToReal(FldVal(base + vsHeatSetPtDesignOff)) .GT. StringToReal(FldVal(base + vsPreheatSetPtConstantOff))) THEN
+    IF((preHeatCoilType .NE. ctNone) .AND. (StringToReal(FldVal(base + vsHeatSetPtDesignOff)) .GT. StringToReal(FldVal(base + vsPreheatSetPtConstantOff)))) THEN
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:10238]: This warning is only useful it there is a preheat coil.

File: src/ExpandObjects/epfilter.f90 [hunk 5]
```diff
@@ -11181,6 +11181,13 @@ SUBROUTINE CreateNewVAV
   !AIR LOOP EQUIPMENT LIST ~ line 552
   CALL CreateNewObj('AirLoopHVAC:OutdoorAirSystem:EquipmentList')
   CALL AddToObjFld('Name', base + vsAirHandlerNameOff,' OA System Equipment')
+  IF (heatRecovery .EQ. htrecSens) THEN
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:11184]: Fix the order of components to match node connections.

File: src/ExpandObjects/epfilter.f90 [hunk 6]
```diff
@@ -11628,9 +11628,9 @@ SUBROUTINE CreateNewVAV
   ! Warn if fixed heating or preheat setpoint is greater than fixed cooling setpoint, and no reset option is active
   IF (isCoolStPtSchedBlank .AND. (coolSetPtReset == csprNone)) THEN
     IF ((heatCoilType .NE. ctNone) .AND. isHeatStPtSchedBlank .AND. (heatSetPtReset == hsprNone)) THEN
-      IF (FldVal(base + pvavsHeatSetPtDesignOff) .GT. FldVal(base + pvavsCoolSetPtDesignOff)) THEN
+      IF (StringToReal(FldVal(base + pvavsHeatSetPtDesignOff)) .GT. StringToReal(FldVal(base + pvavsCoolSetPtDesignOff))) THEN
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:11631]: Fix GT to be on numbers, not strings - duh.

File: src/ExpandObjects/epfilter.f90 [hunk 7]
```diff
@@ -12435,45 +12435,47 @@ SUBROUTINE CreateNewVAV
...
-    CALL AddToObjStr('Humidifying Relative Humidity Setpoint Schedule Name', &
-                     'HVACTemplate-Always ' // TRIM(FldVal(base + pvavsHumidSetPtOff)))
-    CALL AddToObjStr('Dehumidifying Relative Humidity Setpoint Schedule Name', &
-                     'HVACTemplate-Always ' // TRIM(FldVal(base + pvavsDehumSetPtOff)),.TRUE.)
-    CALL AddAlwaysSchedule(FldVal(base + pvavsHumidSetPtOff))
-    CALL AddAlwaysSchedule(FldVal(base + pvavsDehumSetPtOff))
-  ELSE
-    IF (.NOT. isDehumidifyNone) THEN
-      !   Dehumidification humidistat
+  IF (.NOT. isHumidifierNone .OR. .NOT. isDehumidifyNone) THEN
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:12438]: Fix logic for adding humidistat(s) - only if needed.

File: src/ExpandObjects/epfilter.f90 [hunk 8]
```diff
@@ -14391,7 +14393,7 @@ SUBROUTINE CreateNewUnitary
   CALL CreateNewObj('SetpointManager:SingleZone:Cooling')
   CALL AddToObjFld('Name', base + usAirHandlerNameOff,' Economizer Supply Air Temp Manager')
   CALL AddToObjStr('Control Variable','Temperature')
-  CALL AddToObjStr('minimum supply air temperature {C}','13')
+  CALL AddToObjFld('minimum supply air temperature {C}',base + usCoolDesignSupplyTempOff,'')
   CALL AddToObjStr('maximum supply air temperature {C}','45')
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:14397]: This is the cooling setpoint manager, so keep the fixed upper limit. Basically, this allows the cooling setpoint to go high enough to turn off the cooling coil if there's not cooling load.

File: src/ExpandObjects/epfilter.f90 [hunk 9]
```diff
@@ -14426,11 +14428,7 @@ SUBROUTINE CreateNewUnitary
   ENDIF
   CALL AddToObjFld('Economizer Maximum Limit Dry-Bulb Temperature {C}', base + usEconoUpLimitOff,' ')
   CALL AddToObjFld('Economizer Maximum Limit Enthalpy {J/kg}', base + usEconoEnthLimitOff,' ')
-  IF (econoKind .EQ. etFixedDewPointAndDryBulb) THEN
-    CALL AddToObjFld('Economizer Maximum Limit Dewpoint Temperature (C)',base + usEconoUpDewLimitOff, ' ')
-  ELSE
-    CALL AddToObjStr('Economizer Maximum Limit Dewpoint Temperature (C)',' ')
-  ENDIF
+  CALL AddToObjFld('Economizer Maximum Limit Dewpoint Temperature (C)',base + usEconoUpDewLimitOff, ' ')
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:14431]: Dew point limit should always be used if input, regardless of economizer type (per documentation).

File: src/ExpandObjects/epfilter.f90 [hunk 10]
```diff
@@ -15000,7 +14998,7 @@ SUBROUTINE CreateNewUnitary
   CALL AddToObjFld('Availability Schedule Name', base + uhpsHeatAvailSchedNameOff,'')
   CALL AddToObjFld('Rated Total Heating Capacity {W}', base + uhpsHPHeatCoilCapOff,'')
   CALL AddToObjFld('Rated COP', base + uhpsHPHeatCoilCOPOff,'')
-  CALL AddToObjFld('Rated Air Flow Rate {m3/s}', base + uhpsSupplyHeatFlowRateOff,'')
+  CALL AddToObjStr('Rated Air Flow Rate {m3/s}', 'autosize')
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:15001]: Component flow rates/capacities should always be 'autosize'. User values flow into the sizing objects.

File: src/ExpandObjects/epfilter.f90 [hunk 11]
```diff
@@ -22201,6 +22199,13 @@ SUBROUTINE CreateNewDualDuct
   !AIR LOOP EQUIPMENT LIST ~ line 552
   CALL CreateNewObj('AirLoopHVAC:OutdoorAirSystem:EquipmentList')
   CALL AddToObjFld('Name', base + ddsAirHandlerNameOff,' OA System Equipment')
+  IF (heatRecovery .EQ. htrecSens) THEN
```
Comment (mjwitte) [inline src/ExpandObjects/epfilter.f90:22202]: Again - fix order of OA components to match node flow order.

**Review by mjwitte — COMMENTED** [review]
@Myoldmopar Walkthru.

**Comment by Myoldmopar** [comment]
Thanks for the walkthrough, it all makes sense.  This is a highly valuable set of changes, and again, I am really excited about the robustness improvements to ExpandObjects as an unexpected outcome from the Python conversion.  I'm going to merge this in.

**Pull request merged**